### ISSUE TITLE
(?>(?:)...) match failure

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12205]
Migrated from [rt.perl.org#113770](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=113770) (status was 'resolved')

Searchable as RT113770$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



$ perl5\.17\.0 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
OK
$ perl5\.17\.1 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
BUG

This breaks the IPv6 address parsing in Net&#8203;::\{FTP\,HTTP\}&#8203;::Tiny\.  bisect says&#8203;:

3b6759a6b10290b6fea26fb98b50fbf5ce4466b5 is the first bad commit
commit 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Wed Jun 6 08&#8203;:53&#8203;:05 2012 \+0200

&nbsp;    optimise \(?&#8203;:|\) and related NOTHING like constructs out of the compiled optree
&nbsp;    
&nbsp;    A pattern like \(?&#8203;:|\) causes the regex engine to do extra work even though
&nbsp;    it is equivelent to matching nothing\. This optimises such sequences into
&nbsp;    more efficient opcodes that do less work\, and in some cases optimises
&nbsp;    them away entirely\.

:100644 100644 5ec3bc4c85f768245e2c25ca863918b57e2ce1c6 906add73130394ba7f871a2ecfd82e24b5d49348 M      regcomp\.c

I've got some major deja vu about a should\-be\-null \(?&#8203;:\) tickling a
regexp bug that shows up in IPv6 address parsing\, but I can't find the
prior case\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.17.1:

Configured by zefram at Wed Jun 20 23:58:39 BST 2012.

Summary of my perl5 (revision 5 version 17 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=3.2.0-0.bpo.2-amd64, archname=x86_64-linux-thread-multi
    uname='linux barba.rous.org 3.2.0-0.bpo.2-amd64 #1 smp sun jun 3 21:40:57 utc 2012 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/zefram/usr/perl/perl_install/perl-5.17.1-i64-f52 -Duselargefiles -Dusethreads -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dusedevel -Uversiononly -Ui_db'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.5', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.11.3.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.11.3'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/home/zefram/usr/perl/perl_install/perl-5.17.1-i64-f52/lib/5.17.1/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.17.1:
    /home/zefram/usr/perl/perl_install/perl-5.17.1-i64-f52/lib/site_perl/5.17.1/x86_64-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.17.1-i64-f52/lib/site_perl/5.17.1
    /home/zefram/usr/perl/perl_install/perl-5.17.1-i64-f52/lib/5.17.1/x86_64-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.17.1-i64-f52/lib/5.17.1
    .


Environment for perl 5.17.1:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/usr/perl/perl_install/perl-5.17.1-i64-f52/bin:/home/zefram/usr/perl/util:/home/zefram/pub/x86_64-unknown-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

On 21 June 2012 12&#8203;:49\, Zefram \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Zefram
> \# Please include the string&#8203;:  \[perl \#113770\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=113770 >
>
>
>
> This is a bug report for perl from zefram@&#8203;fysh\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.17\.1\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> $ perl5\.17\.0 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
> OK
> $ perl5\.17\.1 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
> BUG
>
> This breaks the IPv6 address parsing in Net&#8203;::\{FTP\,HTTP\}&#8203;::Tiny\.  bisect says&#8203;:
>
> 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5 is the first bad commit
> commit 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5
> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
> Date&#8203;:   Wed Jun 6 08&#8203;:53&#8203;:05 2012 \+0200
>
>    optimise \(?&#8203;:|\) and related NOTHING like constructs out of the compiled optree
>
>    A pattern like \(?&#8203;:|\) causes the regex engine to do extra work even though
>    it is equivelent to matching nothing\. This optimises such sequences into
>    more efficient opcodes that do less work\, and in some cases optimises
>    them away entirely\.
>
> :100644 100644 5ec3bc4c85f768245e2c25ca863918b57e2ce1c6 906add73130394ba7f871a2ecfd82e24b5d49348 M      regcomp\.c
>
> I've got some major deja vu about a should\-be\-null \(?&#8203;:\) tickling a
> regexp bug that shows up in IPv6 address parsing\, but I can't find the
> prior case\.

How incredibly irritating\. I will fix\.

Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 21 June 2012 16&#8203;:13\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> On 21 June 2012 12&#8203;:49\, Zefram \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>> \# New Ticket Created by  Zefram
>> \# Please include the string&#8203;:  \[perl \#113770\]
>> \# in the subject line of all future correspondence about this issue\.
>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=113770 >
>>
>>
>>
>> This is a bug report for perl from zefram@&#8203;fysh\.org\,
>> generated with the help of perlbug 1\.39 running under perl 5\.17\.1\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> $ perl5\.17\.0 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
>> OK
>> $ perl5\.17\.1 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
>> BUG
>>
>> This breaks the IPv6 address parsing in Net&#8203;::\{FTP\,HTTP\}&#8203;::Tiny\.  bisect says&#8203;:
>>
>> 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5 is the first bad commit
>> commit 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5
>> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
>> Date&#8203;:   Wed Jun 6 08&#8203;:53&#8203;:05 2012 \+0200
>>
>>    optimise \(?&#8203;:|\) and related NOTHING like constructs out of the compiled optree
>>
>>    A pattern like \(?&#8203;:|\) causes the regex engine to do extra work even though
>>    it is equivelent to matching nothing\. This optimises such sequences into
>>    more efficient opcodes that do less work\, and in some cases optimises
>>    them away entirely\.
>>
>> :100644 100644 5ec3bc4c85f768245e2c25ca863918b57e2ce1c6 906add73130394ba7f871a2ecfd82e24b5d49348 M      regcomp\.c
>>
>> I've got some major deja vu about a should\-be\-null \(?&#8203;:\) tickling a
>> regexp bug that shows up in IPv6 address parsing\, but I can't find the
>> prior case\.
>
> How incredibly irritating\. I will fix\.

I just pushed yves/rt\_113770 which should fix this bug\. And btw\, I
remember this bug from before as well\.

I will add tests once i get the smoke results \(ill put it as
smoke\-me/yves\-rt\_113770\)

yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On 21 June 2012 17&#8203;:15\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
> On 21 June 2012 16&#8203;:13\, demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
>> On 21 June 2012 12&#8203;:49\, Zefram \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>> \# New Ticket Created by  Zefram
>>> \# Please include the string&#8203;:  \[perl \#113770\]
>>> \# in the subject line of all future correspondence about this issue\.
>>> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=113770 >
>>>
>>>
>>>
>>> This is a bug report for perl from zefram@&#8203;fysh\.org\,
>>> generated with the help of perlbug 1\.39 running under perl 5\.17\.1\.
>>>
>>>
>>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>>> \[Please describe your issue here\]
>>>
>>> $ perl5\.17\.0 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
>>> OK
>>> $ perl5\.17\.1 \-e 'print "\[A\]" =~ /\\A\(?>\\\[\(?&#8203;:\(?&#8203;:\)\(?&#8203;:R\)\{1\}|T|V?|A\)\\\]\)\\z/ ? "OK\\n" : "BUG\\n"'
>>> BUG
>>>
>>> This breaks the IPv6 address parsing in Net&#8203;::\{FTP\,HTTP\}&#8203;::Tiny\.  bisect says&#8203;:
>>>
>>> 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5 is the first bad commit
>>> commit 3b6759a6b10290b6fea26fb98b50fbf5ce4466b5
>>> Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
>>> Date&#8203;:   Wed Jun 6 08&#8203;:53&#8203;:05 2012 \+0200
>>>
>>>    optimise \(?&#8203;:|\) and related NOTHING like constructs out of the compiled optree
>>>
>>>    A pattern like \(?&#8203;:|\) causes the regex engine to do extra work even though
>>>    it is equivelent to matching nothing\. This optimises such sequences into
>>>    more efficient opcodes that do less work\, and in some cases optimises
>>>    them away entirely\.
>>>
>>> :100644 100644 5ec3bc4c85f768245e2c25ca863918b57e2ce1c6 906add73130394ba7f871a2ecfd82e24b5d49348 M      regcomp\.c
>>>
>>> I've got some major deja vu about a should\-be\-null \(?&#8203;:\) tickling a
>>> regexp bug that shows up in IPv6 address parsing\, but I can't find the
>>> prior case\.
>>
>> How incredibly irritating\. I will fix\.
>
> I just pushed yves/rt\_113770 which should fix this bug\. And btw\, I
> remember this bug from before as well\.
>
> I will add tests once i get the smoke results \(ill put it as
> smoke\-me/yves\-rt\_113770\)

I didn't like that patch\, but i have pushed a new version\, with tests\.
If it passes I will rebase it \(and fix the typo in the message\)\.

cheers\,
Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

Resolved with&#8203;:

commit 190c19104d65b8ec0002df12344af2221fa31966
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Thu Jun 21 20&#8203;:42&#8203;:07 2012 \+0200

&nbsp;    turns out we cant maike a jump trie that starts with a NOTHING regop
&nbsp;    
&nbsp;    NOTHING nodes are 1 regop wide\. We need 2 regops worth of
&nbsp;    space\. Which means that we can not safely optimise a sequence
&nbsp;    that starts with a NOTHING regop that needs to turn into
&nbsp;    a jump trie\. We therefore ignore such sequences\.
&nbsp;    
&nbsp;    see perl \#113770 for details\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Resolved with&#8203;:

commit 190c19104d65b8ec0002df12344af2221fa31966
Author&#8203;: Yves Orton \<demerphq@&#8203;gmail\.com>
Date&#8203;:   Thu Jun 21 20&#8203;:42&#8203;:07 2012 \+0200

&nbsp;    turns out we cant maike a jump trie that starts with a NOTHING regop
&nbsp;    
&nbsp;    NOTHING nodes are 1 regop wide\. We need 2 regops worth of
&nbsp;    space\. Which means that we can not safely optimise a sequence
&nbsp;    that starts with a NOTHING regop that needs to turn into
&nbsp;    a jump trie\. We therefore ignore such sequences\.
&nbsp;    
&nbsp;    see perl \#113770 for details\.

**Comment by p5pRT** [comment]
@demerphq - Status changed from 'open' to 'resolved'
